### PR TITLE
reef: osd/OSDMap: Check for uneven weights & != 2 buckets post stretch mode

### PR DESCRIPTION
**Pull request by kamoltat**
Added additional health warnings in
    ``OSDMap::check_health()``.
    
1. Check for uneven weights of the two dividing
buckets after stretch mode is enabled and spit out
warnings if true.
    
2. Check for != 2 buckets after stretch mode is enabled
and spit out warnings if true.

Added standalone test to check to verify the fix.

Fixes: https://tracker.ceph.com/issues/57570

Back porting relevant commits from main PR:

https://github.com/ceph/ceph/pull/48209

Signed-off-by: Kamoltat <ksirivad@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [x] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: doc/rados/operations/health-checks.rst [hunk 1]
```diff
@@ -1406,6 +1406,31 @@ other performance issue with the OSDs.
 The exact size of the snapshot trim queue is reported by the ``snaptrimq_len``
 field of ``ceph pg ls -f json-detail``.
 
+Stretch Mode
+------------
+
+INCORRECT_NUM_BUCKETS_STRETCH_MODE
+__________________________________
+
+Stretch mode currently only support 2 dividing buckets with OSDs, this warning suggests
```
Comment (ronen-fr) [inline doc/rados/operations/health-checks.rst:1415]: supports?

Comment (ronen-fr) [reply doc/rados/operations/health-checks.rst:1415]: do you mean "up to two dividing buckets"?

File: doc/rados/operations/health-checks.rst [hunk 2]
```diff
@@ -1406,6 +1406,31 @@ other performance issue with the OSDs.
...
+
+We encourage you to fix this by removing additional dividing buckets or bump the
+number of dividing buckets to 2.
+
+UNEVEN_WEIGHTS_STRETCH_MODE
+___________________________
+
+The 2 dividing buckets must have equal weights when stretch mode is enabled.
+This warning suggests that the 2 dividing buckets have uneven weights after
+stretch mode is enabled. This is not immediately fatal, however, you can expect
```
Comment (ronen-fr) [inline doc/rados/operations/health-checks.rst:1427]: was?

File: qa/standalone/ceph-helpers.sh [hunk 3]
```diff
@@ -1747,6 +1747,29 @@ function test_wait_for_peered() {
 
 #######################################################################
 
+##
+# Wait until the cluster's health condition disappeared.
```
Comment (ronen-fr) [inline qa/standalone/ceph-helpers.sh:1751]: disappears?

File: qa/standalone/ceph-helpers.sh [hunk 4]
```diff
@@ -1747,6 +1747,29 @@ function test_wait_for_peered() {
...
+# $TIMEOUT default
+#
+# @param string to grep for in health detail
+# @return 0 if the cluster health doesn't matches request,
```
Comment (ronen-fr) [inline qa/standalone/ceph-helpers.sh:1755]: match

File: qa/standalone/ceph-helpers.sh [hunk 5]
```diff
@@ -1747,6 +1747,29 @@ function test_wait_for_peered() {
...
+# 1 otherwise if after $TIMEOUT seconds health condition remains.
+#
+function wait_for_health_gone() {
+    local grepstr=$1
+    local -a delays=($(get_timeout_delays $TIMEOUT .1))
+    local -i loop=0
```
Comment (ronen-fr) [inline qa/standalone/ceph-helpers.sh:1761]: you are not checking whether the string was there on entry, right?

File: src/osd/OSDMap.cc [hunk 6]
```diff
@@ -7201,6 +7201,24 @@ void OSDMap::check_health(CephContext *cct,
 			    ss.str(), 0);
     }
   }
+  // UNEQUAL_WEIGHT
+  if (stretch_mode_enabled) {
+    vector<int> subtrees;
```
Comment (ronen-fr) [inline src/osd/OSDMap.cc:7206]: get_subtree_of_type() is hardly used, and in all placed - the vector is an out-only
parameter. Maybe this is a good opportunity to modify its signature
to
vector<int> get_subtree_of_type(bucket)

File: src/osd/OSDMap.cc [hunk 7]
```diff
@@ -7201,6 +7201,24 @@ void OSDMap::check_health(CephContext *cct,
...
+    crush->get_subtree_of_type(stretch_mode_bucket, &subtrees);
+    if (subtrees.size() != 2) {
+      stringstream ss;
```
Comment (ronen-fr) [inline src/osd/OSDMap.cc:7209]: no need for the stringstream here

File: src/osd/OSDMap.cc [hunk 8]
```diff
@@ -7201,6 +7201,24 @@ void OSDMap::check_health(CephContext *cct,
...
+      ss << "Stretch mode buckets != 2";
+      checks->add("INCORRECT_NUM_BUCKETS_STRETCH_MODE", HEALTH_WARN, ss.str(), 0);
+      return;
+    }
+    int weight1 = crush->get_item_weight(subtrees[0]);
+    int weight2 = crush->get_item_weight(subtrees[1]);
+    stringstream ss;
+    if (weight1 != weight2) {
+      ss << "Stretch mode buckets have different weights!";
```
Comment (ronen-fr) [inline src/osd/OSDMap.cc:7218]: nor here

**Comment by kamoltat** [comment]
@ronen-fr if there is nothing wrong with the backport itself, can you please approve the PR?

**Review by ronen-fr — APPROVED** [review]

**Comment by zdover23** [comment]
> @ronen-fr I appreciate your review, I'll leave it up to the doc writer team to clean up any mistakes I've made with the doc changes. Since this is a backport, I'll note down the change you suggest with the code and will file another separately to address the suggested change.

@kamoltat: I'll take care of the docs matters when this backport is merged.

**Comment by yuriw** [comment]
jenkins test make check

**Comment by kamoltat** [comment]
@zdover23 merged! thank you

**Pull request merged**